From 2505629da55916db8f29a5fc0ba13be77626a01f Mon Sep 17 00:00:00 2001
From: Ciprian Costea <ciprianmarian.costea@nxp.com>
Date: Tue, 22 Nov 2022 13:36:08 +0200
Subject: [PATCH 09/13] hwconfig: fixup: Do not return from
 'ofnode_append_prop' if property does not exist

'lenp' parameter from 'ofnode_get_property' gets set to an error
code if the property does not exist. In our case we would like to
create the property if it does not exist and not directly return.
Otherwise, this will lead to Serdes not being initialized
correctly and PCIe probe will fail.

Issue: ALB-9558
Upstream-Status: Pending 

Signed-off-by: Ciprian Costea <ciprianmarian.costea@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/mach-s32/s32-cc/hwconfig_fixups.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/arch/arm/mach-s32/s32-cc/hwconfig_fixups.c b/arch/arm/mach-s32/s32-cc/hwconfig_fixups.c
index ca07a3aba5..712f1bfe9c 100644
--- a/arch/arm/mach-s32/s32-cc/hwconfig_fixups.c
+++ b/arch/arm/mach-s32/s32-cc/hwconfig_fixups.c
@@ -124,14 +124,9 @@ static int ofnode_append_prop(ofnode node, const char *prop,
 		return -EINVAL;
 
 	old_val = ofnode_get_property(node, prop, &old_len);
-	if (old_len < 0)
-		return -EINVAL;
-
-	if (old_len + len < len || old_len + len < old_len)
-		return -EINVAL;
 
-	/* New property */
-	if (!old_val) {
+	/* Check if new property */
+	if (!old_val || old_len < 0) {
 		new_val = malloc(len);
 		if (!new_val)
 			return -ENOMEM;
@@ -140,6 +135,9 @@ static int ofnode_append_prop(ofnode node, const char *prop,
 		return ofnode_write_prop(node, prop, len, new_val);
 	}
 
+	if (old_len + len < len || old_len + len < old_len)
+		return -EINVAL;
+
 	new_val = malloc(old_len + len);
 	if (!new_val)
 		return -ENOMEM;
-- 
2.17.1

